beacon-chain/db/kv/custody.go

```diff
@@ -0,0 +1,129 @@
+package kv
+
+import (
+	"context"
+
+	"github.com/OffchainLabs/prysm/v6/consensus-types/primitives"
+	"github.com/OffchainLabs/prysm/v6/encoding/bytesutil"
+	"github.com/OffchainLabs/prysm/v6/monitoring/tracing/trace"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	bolt "go.etcd.io/bbolt"
+)
+
+// UpdateCustodyInfo atomically updates the custody group count only it is greater than the stored one.
+// In this case, it also updates the earliest available slot with the provided value.
+// It returns the (potentially updated) custody group count and earliest available slot.
+func (s *Store) UpdateCustodyInfo(ctx context.Context, earliestAvailableSlot primitives.Slot, custodyGroupCount uint64) (primitives.Slot, uint64, error) {
+	_, span := trace.StartSpan(ctx, "BeaconDB.UpdateCustodyInfo")
+	defer span.End()
+
+	storedGroupCount, storedEarliestAvailableSlot := uint64(0), primitives.Slot(0)
+	if err := s.db.Update(func(tx *bolt.Tx) error {
+		// Retrieve the custody bucket.
+		bucket, err := tx.CreateBucketIfNotExists(custodyBucket)
+		if err != nil {
+			return errors.Wrap(err, "create custody bucket")
+		}
+
+		// Retrieve the stored custody group count.
+		storedGroupCountBytes := bucket.Get(groupCountKey)
+		if len(storedGroupCountBytes) != 0 {
+			storedGroupCount = bytesutil.BytesToUint64BigEndian(storedGroupCountBytes)
+		}
+
+		// Retrieve the stored earliest available slot.
+		storedEarliestAvailableSlotBytes := bucket.Get(earliestAvailableSlotKey)
+		if len(storedEarliestAvailableSlotBytes) != 0 {
+			storedEarliestAvailableSlot = primitives.Slot(bytesutil.BytesToUint64BigEndian(storedEarliestAvailableSlotBytes))
+		}
+
+		// Exit early if the new custody group count is lower than or equal to the stored one.
+		if custodyGroupCount <= storedGroupCount {
+			return nil
+		}
+
+		storedGroupCount, storedEarliestAvailableSlot = custodyGroupCount, earliestAvailableSlot
+
+		// Store the earliest available slot.
+		bytes := bytesutil.Uint64ToBytesBigEndian(uint64(earliestAvailableSlot))
+		if err := bucket.Put(earliestAvailableSlotKey, bytes); err != nil {
+			return errors.Wrap(err, "put earliest available slot")
+		}
+
+		// Store the custody group count.
+		bytes = bytesutil.Uint64ToBytesBigEndian(custodyGroupCount)
+		if err := bucket.Put(groupCountKey, bytes); err != nil {
+			return errors.Wrap(err, "put custody group count")
+		}
+
+		return nil
+	}); err != nil {
+		return 0, 0, err
+	}
+
+	log.WithFields(logrus.Fields{
+		"earliestAvailableSlot": storedEarliestAvailableSlot,
+		"groupCount":            storedGroupCount,
+	}).Debug("Custody info")
+
+	return storedEarliestAvailableSlot, storedGroupCount, nil
+}
+
+// UpdateSubscribedToAllDataSubnets updates the "subscribed to all data subnets" status in the database
```

Comment:
the function name + the logic inside make it feel like it's doing a lot more than just storing a true value, i wonder what we can do to simplify

Reply:
The logic inside is really what is described by the godoc.
I agree the function name could be changed if you have a better idea.

> i wonder what we can do to simplify

The function has to do what the godoc states, in a atomic way. If you find a simpler way to do it, I'm open to it!